### PR TITLE
server: Always respond to getheaders.

### PR DESCRIPTION
**Pull request by davecgh**
The existing logic for responding to `getheaders` is based on the old sync model before headers-first syncing was introduced and thus avoids responding whenever the current local chain is not considered to be current.  This existing behavior was important under the the old sync model since it would otherwise lead to unnecessarily downloading a bunch of blocks among other undesirable corner cases, however, it does come with some downsides.

For example, one less than ideal consequence of not responding is that it can lead to peers appearing to be unresponsive and/or stalled rather than merely not having anything interesting yet.

Also, there are a variety of corner cases where a peer might temporarily no longer consider itself current such as after being unable to communicate with the network for a long time, or in testing scenarios where there are necessarily long periods of time without any new blocks.

However, since peers now use the headers to determine what blocks to download, serving up headers before the peer is known to be current no longer leads to undesirable behavior.

Taking the aforementioned factors into consideration, this modifies the semantics to always respond to `getheaders` requests, even before the local chain is fully synced.

Further, in order to avoid needlessly sending extra data early during the initial sync process and to provide some additional protection against low-effort DoS attempts, a new check is introduced to respond with empty headers when the local chain does not yet have the minimum cumulative work value already known from the chain parameters to have been achieved on the network.  Responding with empty headers signals to the remote peer that there are not any interesting headers available without the local peer appearing unresponsive.

### CODE REVIEW

File: server.go [hunk 1]
```diff
@@ -471,6 +472,13 @@ type server struct {
 	bytesSent     uint64 // Total bytes sent by all peers since start.
 	shutdown      int32
 
+	// minKnownWork houses the minimum known work from the associated network
+	// params converted to a uint256 so the conversion only needs to be
+	// performed once when the server is initialized.  Ideally, the chain params
```
Comment (ukane-philemon) [inline server.go:477]: nit: rm extra space here

File: server.go [hunk 2]
```diff
@@ -1258,13 +1266,18 @@ func (sp *serverPeer) OnGetBlocks(_ *peer.Peer, msg *wire.MsgGetBlocks) {
...
 	chain := sp.server.chain
-	if !sp.server.syncManager.IsCurrent() && (len(locatorHashes) == 0 ||
-		*locatorHashes[0] != chain.BestSnapshot().PrevHash) {
-
+	tipHash := chain.BestSnapshot().Hash
+	workSum, err := chain.ChainWork(&tipHash)
+	if err == nil && workSum.Lt(&sp.server.minKnownWork) {
+		srvrLog.Debugf("Sending empty headers to peer %s in response to "+
+			"getheaders due to local best knwon tip having too little work",
+			sp.Peer)
```
Comment (ukane-philemon) [inline server.go:1279]: nit: Is the `%s` directive appropriate for `sp.Peer`? Maybe you mean `sp.Peer.String()`? or use the `%v` directive.

**Review by matheusd — APPROVED** [review]
Tested by having a mainnet node sync, then connecting a second instance to this syncing node.

With this PR applied, I get the expected result (second node starts syncing headers, then blocks once the first node reaches the minKnownWork height):

```
# On the second node
2022-11-30 09:54:04.256 [INF] SYNC: Syncing headers to block height 611166 from peer 127.0.0.1:9108 (outbound)         
2022-11-30 09:54:04.341 [INF] SYNC: Processed 2000 headers in the last 75.11s (progress 0.28%)
2022-11-30 09:54:14.404 [INF] SYNC: Processed 210000 headers in the last 10.06s (progress 29.57%)               
2022-11-30 09:54:24.512 [INF] SYNC: Processed 162000 headers in the last 10.11s (progress 52.13%)                                              
2022-11-30 09:54:34.557 [INF] SYNC: Processed 154000 headers in the last 10.04s (progress 73.63%)                                              
2022-11-30 09:54:39.803 [INF] SYNC: Processed 86000 headers in the last 5.25s (progress 85.38%)
2022-11-30 09:54:39.803 [INF] SYNC: Initial headers sync complete (best header hash 000000000000000001cb6b6c069aa1b1fd16a678e95a91fb595f5df3855
a3a3b, height 612000)                                                                                                                          
```

I also tested dcrwallet in spv mode connecting to the updated dcrd node, and it seems to be working correctly (I'm triggering an error while syncing but I believe it's unrelated to this PR). For reference, the dcrwallet error reported with an empty `getheaders` response is the following:

```
2022-11-30 09:44:35.560 [WRN] SYNC: Lost peer 127.0.0.1:9108: protocol violation: peer did not provide headers through advertised height
```

File: server.go [hunk 3]
```diff
@@ -3406,6 +3420,15 @@ func newServer(ctx context.Context, listenAddrs []string, db database.DB,
 		quit:            make(chan struct{}),
 	}
 
+	// Convert the minimum known work to a uint256 when it exists.  Ideally, the
```
Comment (ukane-philemon) [inline server.go:3423]: nit: rm extra space.

**Review by ukane-philemon — CHANGES_REQUESTED** [review]

File: server.go (hunk 1 above)
Comment (davecgh) [reply server.go:477]: Pretty much all comments in dcrd use double spaces after periods.  This is one of those things like 80 cols that can be a point of contention.  See https://link.springer.com/article/10.3758/s13414-018-1527-6.

File: server.go (hunk 2 above)
Comment (davecgh) [reply server.go:1279]: Yep, `%s` is intended as it is the specific formatting directive for strings and `sp.Peer` implements [fmt.Stringer](https://pkg.go.dev/fmt#Stringer).

File: server.go [hunk 4]
```diff
@@ -1258,13 +1266,18 @@ func (sp *serverPeer) OnGetBlocks(_ *peer.Peer, msg *wire.MsgGetBlocks) {
 
 // OnGetHeaders is invoked when a peer receives a getheaders wire message.
 func (sp *serverPeer) OnGetHeaders(_ *peer.Peer, msg *wire.MsgGetHeaders) {
-	// Ignore getheaders requests if not in sync unless the local best chain
-	// is exactly at the same tip as the requesting peer.
-	locatorHashes := msg.BlockLocatorHashes
+	// Send an empty headers message in the case the local best known chain does
+	// not have the minimum cumulative work value already known to have been
+	// achieved on the network.  This signals to the remote peer that there are
```
[the remark this answers is not in the document]
Comment (davecgh) [reply server.go:1271]: Intentional.  See previous response.

File: server.go (hunk 3 above)
Comment (davecgh) [reply server.go:3423]: Intentional.  See previous response.

File: server.go [hunk 5]
```diff
@@ -1258,13 +1266,18 @@ func (sp *serverPeer) OnGetBlocks(_ *peer.Peer, msg *wire.MsgGetBlocks) {
...
+	// not any interesting headers available without appearing unresponsive.
```
Comment (dnldd) [inline server.go:1272]: consider changing `...that there are not any...` to `...that there are no...`. The former is correct, the later reads better I think.

**Comment by davecgh** [comment]
> Ah, the dcrwallet issue is due to `OnGetCFilterV2` enforcing the `IsCurrent()` check without taking into account the `minKnownWork` exception introduced by this PR.
> 
> I believe you should also change that handler to use the same check (or abstract the checks performed in both cases to a common function such as `canProvideBlocks()` or something to avoid this issue in the future)

I'll open a separate PR to address this one.  I'll reason through it a bit more, but I highly suspect it probably makes sense to remove the check entirely for that path now because it was really just blindly inherited from the old bloom filter days where it mattered because bloom filters put extra load on the server.  That is no longer the case for version 2 cfilters.

File: server.go (hunk 5 above)
Comment (davecgh) [reply server.go]: Updated to the suggested variant. I don't have a strong preference either way.

**Review by dnldd — APPROVED** [review]

**Pull request merged**